### PR TITLE
Implement support for arbitrary joypads

### LINKED ISSUE #7071: Allow creation of arbitrary joypads
### Describe the project you are working on

A project supporting the GameCube Controller Adapter

### Describe the problem or limitation you are having in your project

The only way to support the GameCube Controller Adapter in Godot is either by activating some sort of native PC mode (not all adapters support this) or by intercepting the adapter via low-level methods. As the former cannot be relied on, the latter is what I've chosen to utilize via GDExtension & libusb

While this method works, it unfortunately has no means of linking with the Input system to send out native events. I could send out custom events or bind to a dedicated signal from my GDExtension, but it's a headache to need a specialized system to recognize inputs for exactly one input source. Should I decide to add more custom inputs in the future, even more custom events/signals would have to be created and accounted for

This would be possible to work around if I decided to transition from GDExtension to a module, as the methods necessary to add/remove inputs are readily accessable. But as this is something I'd like to release as an Asset Library, a module isn't a feasible route to take. It's also not feasible to simply bind the existing methods, as they rely on internal logic that very much should remain unbound

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Instead of binding the existing methods that require jumping through prerequisite internal hoops, create new methods that handle those hoops safely. To make it clear when these new methods are used, mark joypads created by them as "arbitrary" & only allow modification of joypads marked as such

This would preserve normal joypad logic, while giving a means of creating custom joypads and passing on parsed inputs to Godot's input system. This allows custom joypad logic to be entirely seperate, and a user would only need to drop in an Asset Library supporting it to handle arbitrary joypads as if they were any other joypad

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

See accompanying PR: https://github.com/godotengine/godot/pull/78141/

### If this enhancement will not be used often, can it be worked around with a few lines of script?

If and only if implemented via module, otherwise impossible

### Is there a reason why this should be core and not an add-on in the asset library?

This opens the door for add-ons of custom inputs to be possible in the first place; no custom inputs would actually be added to core

### PR DESCRIPTION
**Pull request by Repiteo**
This PR implements methods for adding/removing gamepads arbitrarily. This is similar to the functionality modules have to setup their own joypads, but adds an extra precautionary layer of marking these joypads as "arbitrary". Thanks to this safeguard, these new methods are binded & can be utilized in GDExtension or even GDScript

Bugsquad edit: closes godotengine/godot-proposals#7071

### CODE REVIEW

**Comment by Calinou** [comment]
Thanks for opening a pull request :slightly_smiling_face: 

Feature pull requests should be associated to a *feature proposal* to justify the need for implementing the feature in Godot core. Please open a proposal on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository (and link to this pull request in the proposal's body).

**Comment by AThousandShips** [comment]
You will have to run `--doctool` on the compiled version of your code to update the documentation

**Comment by Repiteo** [comment]
Added a minor change to `get_unused_joy_id`: if an "arbitrary" bool is passed, it'll now seek inputs from the range 16-255. This will prevent an edgecase where no more traditional inputs could be added if there's too many arbitrary inputs, as there's no longer any risk of overlap

**Comment by Repiteo** [comment]
This most recent rebase required resolving a conflict, and a check I've added arguably overlaps with the newly added `p_joypad_info`. I left the implementation as-is for now, but was wondering if it'd be better to turn the dedicated `p_arbitrary` boolean variable into a dictionary key instead

File: doc/classes/Input.xml [hunk 1]
```diff
@@ -30,6 +30,13 @@
 				If the specified action is already pressed, this will release it.
 			</description>
 		</method>
+		<method name="add_arbitrary_joy">
+			<return type="int" />
+			<param index="0" name="name" type="String" />
+			<description>
+				Returns the device index of a newly created arbitrary joypad.
```
Comment (Calinou) [inline doc/classes/Input.xml:37]: The description should mention what arbitrary joypads can be used for, and how to set them up from a script or extension. You could link to a third-party resource if needed, but it's important not to leave people "in the blind" so to speak :slightly_smiling_face:

Comment (Repiteo) [reply doc/classes/Input.xml:37]: I think I've got a satisfactory implementation. It doesn't include *where* to get the processed data from, but I think that's a bit outside the scope of what the example needs to show

Comment (Repiteo) [reply doc/classes/Input.xml:37]: Added a CSharp section to the example!

File: doc/classes/Input.xml [hunk 2]
```diff
@@ -30,6 +30,58 @@
...
+				    Input.set_arbitrary_joy_button(_index, JOY_BUTTON_B, bytes[3] &gt; 0)
+
+				func _notification(what):
+				    if what == NOTIFICATION_PREDELETE:
+				        Input.remove_arbitrary_joy(_index)
+				[/gdscript]
+				[csharp]
+				private int _index;
+
+				public ExampleClass(string name = "")
```
[the remark this answers is not in the document]
Comment (Repiteo) [reply doc/classes/Input.xml:61]: There doesn't seem to be a 100% consistent implementation style for these codeblocks, so I wasn't sure whether or not class declaration was necessary. Though considering it does make an override, I should either make the class inheritance explicit or change the example to something more generic; I'm leaning towards the latter

[the remark this answers is not in the document]
Comment (Repiteo) [reply doc/classes/Input.xml]: Went with the latter!

File: doc/classes/Input.xml [hunk 3]
```diff
@@ -30,6 +30,68 @@
...
+
+				func destroy_joy():
+				    if _index != -1:
+				        Input.remove_arbitrary_joy(_index)
+				        _index = -1
+				[/gdscript]
+				[csharp]
+				private int _index = -1;
+
+				public CreateJoy(string name = "")
```
Comment (raulsntos) [inline doc/classes/Input.xml:64]: I'm assuming you meant for this to be a void returning method.

```suggestion
				public void CreateJoy(string name = "")
```

Comment (Repiteo) [reply doc/classes/Input.xml]: Forgot to add that after changing it from a constructor, good catch!

**Pull request closed without merging**